File: frontend/src/types/pumps.ts
// export interface PumpDetailIn {
//   doc_customer: string;
//   doc_no: string;
//   doc_date: string;
//   brand: string;
//   model: string;
//   tag_no: string;
//   pump_type_id: string;
//   pump_type_name: string;
//   serial_no: string;
//   media: string;
//   pump_standard_id: string;
//   pump_standard: string;
//   pump_design: string;
//   stage?: number;
//   location?: string;
//   design_temp: string;
//   max_temp: string;
//   solid_type?: string;
//   solid_diameter?: string;
//   density: number;
//   density_unit: string;
//   viscosity: number;
//   viscosity_unit: string;
//   max_flow: number;
//   max_flow_unit: string;
//   min_flow: number;
//   min_flow_unit: string;
//   vapor_pressure: number;
//   vapor_pressure_unit: string;
//   pump_speed: number;
//   pump_speed_unit: string;
//   design_flow: string;
//   design_flow_unit: string;
//   design_head: string;
//   design_head_unit: string;
//   min_head: string;
//   min_head_unit: string;
//   max_head: string;
//   max_head_unit: string;
//   suction_velo: string;
//   suction_velo_unit: string;
//   discharge_velo: string;
//   discharge_velo_unit: string;
//   bep_head: number;
//   bep_head_unit: string;
//   bep_flow: string;
//   bep_flow_unit: string;
//   npshr: string;
//   npshr_unit: string;
//   pump_efficiency: string;
//   pump_efficiency_unit: string;
//   hyd_power: number;
//   hyd_power_unit: string;
//   voltage: string;
//   voltage_unit: string;
//   power_required_cal: string;
//   power_required_cal_unit: string;
//   power_min_flow: number;
//   power_min_flow_unit: string;
//   power_max_flow: number;
//   power_max_flow_unit: string;
//   power_bep_flow: number;
//   power_bep_flow_unit: string;
//   suggest_motor: number;
//   pump_suction_size_id?: string;
//   pump_suction_size?: string;
//   pump_suction_rating_id?: string;
//   pump_suction_rating?: string;
//   pump_discharge_size_id?: string;
//   pump_discharge_size?: string;
//   pump_discharge_rating_id?: string;
//   pump_discharge_rating?: string;
//   suction_pipe_data_id: string;
//   suction_pipe_size: string;
//   suction_pipe_rating?: string;
//   suction_pipe_sch: string;
//   discharge_pipe_data_id: string;
//   discharge_pipe_size: string;
//   discharge_pipe_rating?: string;
//   discharge_pipe_sch: string;
//   tank_position?: string;
//   tank_design?: string;
//   tank_pressure?: number;
//   suction_head?: number;
//   suction_pipe_length?: number;
//   discharge_pipe_length_h?: number;
//   discharge_pipe_length_v?: number;
//   suction_pipe_id: number;
//   discharge_pipe_id: string;
//   suction_elbow?: number;
//   suction_tee?: number;
//   suction_reducer?: number;
//   suction_valve?: number;
//   suction_y_strainer?: number;
//   suction_other?: number;
//   suction_equi_length?: number;
//   discharge_equi_length?: number;
//   discharge_elbow?: number;
//   discharge_tee?: number;
//   discharge_reducer?: number;
//   discharge_valve?: number;
//   discharge_y_strainer?: number;
//   discharge_other?: number;
//   casing_mat_id?: number;
//   casing_mat?: string;
//   shaft_mat_id?: number;
//   shaft_mat?: string;
//   diffuser_mat_id?: number;
//   diffuser_mat?: number;
//   impeller_type_id?: number;
//   impeller_type?: string;
//   design_impeller_dia: number;
//   impeller_max?: number;
//   impeller_mat_id?: number;
//   impeller_mat?: string;
//   materials_wear_ring_id?: number;
//   materials_wear_ring?: string;
//   bearing_nde?: string;
//   bearing_num?: string;
//   bearing_lubric_type?: string;
//   bearing_lubric_brand?: string;
//   bearing_lubric_no?: string;
//   rotation_de_id: number;
//   rotation_de: string;
//   bearing_de?: string;
//   bearing_de_no?: string;
//   bearing_last_chg_dt?: Date;
//   sleeve_mat_id?: number;
//   sleeve_mat?: string;
//   bearing_housing_mat_id?: number;
//   bearing_housing_mat?: string;
//   gland_mat?: string;
//   casing_gas?: string;
//   oring_gas?: number;
//   impeller_gas?: string;
//   pump_lining_mat?: string;
//   base_plate?: string;
//   coup_model?: string;
//   coup_brand?: string;
//   coup_type: string;
//   coup_size?: string;
//   coup_spacer?: number;
//   motor_brand?: string;
//   motor_model?: string;
//   motor_serial_no?: string;
//   motor_drive_id?: number;
//   motor_drive?: string;
//   motor_frame?: string;
//   motor_protection?: string;
//   motor_standard?: string;
//   motor_ie?: string;
//   motor_speed: number;
//   motor_speed_unit: string;
//   motor_rated: number;
//   motor_rated_unit: string;
//   motor_factor: number;
//   motor_connection?: string;
//   motor_phase: number;
//   motor_efficiency: string;
//   motor_efficiency_unit: string;
//   mech_api_id: string;
//   mech_api_plan?: string;
//   mech_main_temp?: number;
//   mech_main_pre?: number;
//   mech_seal_cham?: string;
//   mech_brand?: string;
//   mech_model?: string;
//   mech_size?: number;
//   mech_size_unit?: string;
//   mech_design_id?: number;
//   mech_design?: string;
//   mech_material?: string;
//   concentration?: number;
//   pump_status: string;
// }

export interface PumpDetailOut {
  pump_id: number;
  design_temp: string;
  max_temp: string;
  density: number;
  density_unit: string;
  max_flow: number;
  max_flow_unit: string;
  min_flow: number;
  min_flow_unit: string;
  suction_velo: string;
  suction_velo_unit: string;
  discharge_velo: string;
  discharge_velo_unit: string;
  bep_head: number;
  bep_head_unit: string;
  bep_flow: string;
  bep_flow_unit: string;
  hyd_power: number;
  hyd_power_unit: string;
  voltage: string;
  voltage_unit: string;
  power_required_cal: string;
  power_required_cal_unit: string;
  power_min_flow: number;
  power_min_flow_unit: string;
  power_max_flow: number;
  power_max_flow_unit: string;
  power_bep_flow: number;
  power_bep_flow_unit: string;
  suggest_motor: number;
  tank_position?: string;
  tank_design?: string;
  tank_pressure?: number;
  suction_head?: number;
  suction_pipe_length?: number;
  discharge_pipe_length_h?: number;
  discharge_pipe_length_v?: number;
  suction_pipe_id: number;
  discharge_pipe_id: string;
  suction_elbow?: number;
  suction_tee?: number;
  suction_reducer?: number;
  suction_valve?: number;
  suction_y_strainer?: number;
  suction_other?: number;
  suction_equi_length?: number;
  discharge_equi_length?: number;
  discharge_elbow?: number;
  discharge_tee?: number;
  discharge_reducer?: number;
  discharge_valve?: number;
  discharge_y_strainer?: number;
  discharge_other?: number;
  diffuser_mat_id?: number;
  diffuser_mat?: number;
  materials_wear_ring_id?: number;
  materials_wear_ring?: string;
  sleeve_mat_id?: number;
  sleeve_mat?: string;
  bearing_housing_mat_id?: number;
  bearing_housing_mat?: string;
  gland_mat?: string;
  casing_gas?: string;
  oring_gas?: number;
  impeller_gas?: string;
  pump_lining_mat?: string;
  base_plate?: string;
  concentration?: number;
  pump_status: string;
}

export interface LOVOut {
  data_type: string;
  data_name: string;
  data_value: string;
  additional_1?: string;
  additional_2?: string;
  additional_3?: string;
  created_at: string;
  created_by: string;
  updated_at: string;
  updated_by: string;
}

File: backend/pumps/schemas/pumps.py
from datetime import datetime
from ninja import Schema
from ninja.orm import create_schema

from pumps.models import EngineeringDetail

EngineeringDetailIn = create_schema(model=EngineeringDetail)
EngineeringDetailOut = create_schema(model=EngineeringDetail)

class PumpCustomerDetail(Schema):
    doc_customer: str
    doc_no: str
    doc_date: str

class PumpGeneralDetail(Schema):
    location: str = None
    brand: str
    model: str
    tag_no: str
    serial_no: str
    pump_standard_id: str
    pump_standard: str
    pump_type_id: str
    pump_type_name: str
    stage: int = None
    pump_design: str

class PumpOtherGeneralDetail(Schema):
    design_temp: str
    max_temp: str
    density: float
    density_unit: str
    max_flow: float
    max_flow_unit: str
    min_flow: float
    min_flow_unit: str
    suction_velo: str
    suction_velo_unit: str
    discharge_velo: str
    discharge_velo_unit: str
    bep_head: float
    bep_head_unit: str
    bep_flow: str
    bep_flow_unit: str
    hyd_power: float
    hyd_power_unit: str
    voltage: str
    voltage_unit: str
    power_required_cal: str
    power_required_cal_unit: str
    power_min_flow: float
    power_min_flow_unit: str
    power_max_flow: float
    power_max_flow_unit: str
    power_bep_flow: float
    power_bep_flow_unit: str
    suggest_motor: float
    tank_position: str = None
    tank_design: str = None
    tank_pressure: float = None
    suction_head: float = None
    suction_pipe_length: float = None
    discharge_pipe_length_h: float = None
    discharge_pipe_length_v: float = None
    suction_pipe_id: int
    discharge_pipe_id: str
    suction_elbow: int = None
    suction_tee: int = None
    suction_reducer: int = None
    suction_valve: int = None
    suction_y_strainer: int = None
    suction_other: int = None
    suction_equi_length: float = None
    discharge_equi_length: float = None
    discharge_elbow: int = None
    discharge_tee: int = None
    discharge_reducer: int = None
    discharge_valve: int = None
    discharge_y_strainer: int = None
    discharge_other: int = None
    diffuser_mat_id: int = None
    diffuser_mat: int = None
    materials_wear_ring_id: int = None
    materials_wear_ring: str = None
    sleeve_mat_id: int = None
    sleeve_mat: str = None
    bearing_housing_mat_id: int = None
    bearing_housing_mat: str = None
    gland_mat: str = None
    casing_gas: str = None
    oring_gas: float = None
    impeller_gas: str = None
    pump_lining_mat: str = None
    base_plate: str = None
    concentration: float = None
    pump_status: str

class PumpMaterialDetail(Schema):
    casing_mat_id: int = None
    casing_mat: str = None
    impeller_mat_id: int = None
    impeller_mat: str = None
    shaft_mat_id: int = None
    shaft_mat: str = None

class PumpApplicationData(Schema):
    media: str
    viscosity: float
    viscosity_unit: str
    vapor_pressure: float
    vapor_pressure_unit: str
    solid_type: str = None
    solid_diameter: str = None

class PumpMotorGeneralDetail(Schema):
    motor_brand: str = None
    motor_model: str = None
    motor_serial_no: str = None
    motor_drive_id: int = None
    motor_drive: str = None
    motor_frame: str = None
    motor_protection: str = None
    motor_standard: str = None
    motor_ie: str = None
    motor_speed: float
    motor_speed_unit: str
    motor_rated: float
    motor_rated_unit: str
    motor_factor: float
    motor_connection: str = None
    motor_phase: int
    motor_efficiency: str
    motor_efficiency_unit: str

class PumpCouplingDetails(Schema):
    coup_model: str = None
    coup_brand: str = None
    coup_type: str
    coup_size: str = None
    coup_spacer: float = None

class PumpImpellerDetails(Schema):
    impeller_type_id: int = None
    # impeller_type: str = None
    design_impeller_dia: float
    impeller_max: float = None

class PumpTechnicalData(Schema):
    pump_speed: float
    pump_speed_unit: str
    pump_efficiency: str
    pump_efficiency_unit: str
    min_head: str
    min_head_unit: str
    max_head: str
    max_head_unit: str
    npshr: str
    npshr_unit: str
    design_flow: str
    design_flow_unit: str
    design_head: str
    design_head_unit: str

class PumpMechanicalSealGeneralDetails(Schema):
    mech_seal_cham: str = None

class PumpMechanicalSealDetails(Schema):
    mech_api_id: str
    mech_api_plan: str = None
    mech_main_temp: float = None
    mech_main_pre: float = None
    mech_brand: str = None
    mech_model: str = None
    mech_size: float = None
    mech_size_unit: str = None
    mech_design_id: int = None
    mech_design: str = None
    mech_material: str = None

class PumpFlangeDetails(Schema):
    pump_suction_size_id: str = None
    pump_suction_size: str = None
    pump_suction_rating_id: str = None
    pump_suction_rating: str = None
    pump_discharge_size_id: str = None
    pump_discharge_size: str = None
    pump_discharge_rating_id: str = None
    pump_discharge_rating: str = None
    suction_pipe_data_id: str
    suction_pipe_size: str
    suction_pipe_rating: str = None
    suction_pipe_sch: str
    discharge_pipe_data_id: str
    discharge_pipe_size: str
    discharge_pipe_rating: str = None
    discharge_pipe_sch: str

class PumpBearingDetails(Schema):
    rotation_de_id: int
    rotation_de: str
    bearing_nde: str = None
    bearing_num: str = None
    bearing_lubric_type: str = None
    bearing_lubric_brand: str = None
    bearing_lubric_no: str = None
    bearing_de: str = None
    bearing_de_no: str = None
    bearing_last_chg_dt: datetime = None

class PumpDetailIn(PumpCustomerDetail, PumpGeneralDetail, PumpOtherGeneralDetail, PumpMaterialDetail,
                   PumpApplicationData, PumpMotorGeneralDetail, PumpCouplingDetails, 
                   PumpImpellerDetails, PumpTechnicalData, PumpMechanicalSealGeneralDetails,
                   PumpMechanicalSealDetails, PumpFlangeDetails, PumpBearingDetails):
    """
    Combined schema for all pump details (except pump_id, it's scheme for new data).
    Inherits fields from multiple schemas to create a comprehensive input schema.
    """
    pass

class PumpDetailOut(PumpDetailIn):
    """
    Combined schema for all pump details.
    """
    pump_id: int